src/rescharstring.rs

```diff
@@ -0,0 +1,151 @@
+//!
+//! This is an internal module. It’s public types are re-exported by the
+//! parent.
+
+use std::marker::PhantomData;
+use super::{decode, encode};
+use super::OctetString;
+use super::tag::Tag;
+
+
+//------------ LimitedCharSet ------------------------------------------------
+
+/// Trait for the various Restricted Character String types defined in X.680
+pub trait LimitedCharSet {
+
+    /// Implementations must specify their own natural tag
+    const TAG: Tag;
+
+    /// Implementations must verify the contents
+    fn is_allowed<I: Iterator<Item=u8>>(i: &mut I) -> bool;
+}
+
+#[derive(Debug)]
+pub struct CharSetError;
+
+
+//------------ PrintableStringSet --------------------------------------------
+
+/// PrintableString
+///
+/// The printable string is an implementation of the restricted character
+/// string type defined in X.680 that only allows the following characters:
+/// a-z A-Z 0-9 (space) ' ( ) + , _ . / : = ?
+pub type PrintableString = ResCharString<PrintableStringSet>;
+
+#[derive(Debug)]
+pub struct PrintableStringSet;
+
+impl LimitedCharSet for PrintableStringSet {
+
+    const TAG: Tag = Tag::PRINTABLE_STRING;
+
+    fn is_allowed<I: Iterator<Item=u8>>(i: &mut I) -> bool {
+        i.all(|x| {
+                x.is_ascii_alphanumeric() || // A-Z a-z 0-9
+                x == b' ' || x == b'\'' || x == b'(' || x == b')' ||
+                x == b'+' || x == b',' || x == b'-' || x == b'.' ||
+                x == b'/' || x == b':' || x == b'=' || x == b'?'
+        })
+    }
+}
+
+
+//------------ ResCharString -------------------------------------------------
+
+/// A generic Restricted Character String
+///
+/// Contains an OCTET STRING with the actual content.
+///
+/// Specific implementations have to ensure that they verify the content
+/// according to the appropriate limitations, and use the correct Tag.
+///
+#[derive(Clone, Debug)]
+pub struct ResCharString<L: LimitedCharSet> {
+    octets: OctetString,
+    marker: PhantomData<L>
+}
+
+impl<L: LimitedCharSet> ResCharString<L> {
+
+    unsafe fn new_unchecked(octets: OctetString) -> Self {
+        ResCharString {
+            octets,
+            marker: PhantomData
+        }
+    }
+
+    pub fn new(os: OctetString) -> Result<Self, CharSetError> {
+        if L::is_allowed(&mut os.octets()) {
+            Ok(unsafe { Self::new_unchecked(os) })
+        } else {
+            Err(CharSetError)
+        }
+    }
+
+    pub fn take_from<S: decode::Source>(
+        cons: &mut decode::Constructed<S>
+    ) -> Result<Self, S::Err> {
+        cons.take_value_if(L::TAG, Self::take_content_from)
+    }
+
+    pub fn take_content_from<S: decode::Source>(
+        content: &mut decode::Content<S>
+    ) -> Result<Self, S::Err> {
+        let os = OctetString::take_content_from(content)?;
+        Self::new(os).map_err(|_| decode::Error::Malformed.into())
+    }
+
+    pub fn encode<'a>(&'a self) -> impl encode::Values + 'a {
```

Comment:
We also need `pub fn encode_as<'a>(&'a self, tag: Tag) -> impl encode::Values + 'a`.

Reply:
Done